File: week1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int checkSum(long card_num);

int main(void)
{
    // prompts user for card number
    long number = get_long("Number: ");

    // stores returned checksum value
    int check = checkSum(number);

    // if card is valid then prints type of card
    if (((number / 10000000000000) == 37 || (number / 10000000000000) == 34) && (check % 10 == 0))
    {
        printf("AMEX\n");
    }
    else if (((number / 100000000000000) == 51 || (number / 100000000000000) == 52 || (number / 100000000000000) == 53 ||
              (number / 100000000000000) == 54 || (number / 100000000000000) == 55) &&
             (check % 10 == 0))
    {
        printf("MASTERCARD\n");
    }
    else if (((number / 1000000000000000) == 4 || (number / 1000000000000) == 4) && (check % 10 == 0))
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

// checks the card number and returns checkSum
int checkSum(long card_num)
{
    int checksum, second_last_digit, last_digit, product;
    int sum_of_product = 0;
    int sum_of_leftover_digits = 0;

    while (card_num != 0)
    {
        // last and second last digit
        second_last_digit = (card_num / 10) % 10;
        last_digit = card_num % 10;

        // multiply second last digit with 2 and sum all the products
        product = 2 * second_last_digit;
        if (product >= 10)
        {
            product = (product / 10) + (product % 10);
            sum_of_product += product;
        }
        else
        {
            sum_of_product += product;
        }

        // sum of remaining digits which are multiplied by 2
        sum_of_leftover_digits += last_digit;

        card_num = card_num / 100;
    }

    // checksum
    checksum = sum_of_product + sum_of_leftover_digits;
    return checksum;
}

File: week6/sentimental-credit/credit.py
from cs50 import get_int
import re

# checks for card number is valid or not via Luhn's Algorithm


def check(cardNum):
    sumOfProduct = 0
    sumOfLeftDigits = 0

    while cardNum != 0:
        # last & second last digits of the card
        lastDigit = cardNum % 10
        secondLast = (cardNum // 10) % 10

        # multiply second last digit with 2
        product = secondLast * 2

        # check if product is greater than 10 then make it less than 10
        if product >= 10:
            product = (product // 10) + (product % 10)

        # sum of products of second last and it's consecutive digits
        sumOfProduct += product

        # sum of remaining digits
        sumOfLeftDigits += lastDigit
        cardNum //= 100

    # calculates checksum and returns it
    checkSum = sumOfProduct + sumOfLeftDigits
    return checkSum


# get user input
number = get_int("Number: ")

# stores checksum returned from function
checksum = check(number)

# convert card number to string to use RegEx
card = str(number)

# if last digit of checksum is equals to 0
if checksum % 10 == 0:
    # if card number starts with '34' or '37' and have total 15 digits
    if re.search("^3[4, 7].{13}", card):
        print("AMEX")

    # if card number starts with '51' or '52' or '53' or '54' or '55' and have total 16 digits
    elif re.search("^5[1, 2, 3, 4, 5].{14}", card):
        print("MASTERCARD")

    # if card number starts with '4' and have total 13 or 16 digits
    elif re.search("^4.{12}", card) or re.search("^4.{15}", card):
        print("VISA")

    # if card don't match above requirements
    else:
        print("INVALID")
else:
    print("INVALID")
